refined ceph mds procedure (#640)
openstack-k8s-operators/data-plane-adoption: docs_user/modules/proc_migrating-ceph-mds.adoc

= Migrating {Ceph} MDS to new nodes within the existing cluster

> The following procedure describes how to migrate the MDS daemon when {rhos_component_storage_file_first_ref}, deployed with either a cephfs-native or ceph-nfs back end, is part of the overcloud deployment. The MDS migration is performed by `cephadm`, and you move the daemons placement from a hosts-based approach to a label-based approach.

Reviewer: ```suggestion
You can migrate the MDS daemon when {rhos_component_storage_file_first_ref}, deployed with either a cephfs-native or ceph-nfs back end, is part of the overcloud deployment. The MDS migration is performed by `cephadm`, and you move the daemons placement from a hosts-based approach to a label-based approach.
```

Recommended to remove self-reference

ifeval::["{build}" != "upstream"]
This ensures that you can visualize the status of the cluster and where daemons are placed by using the `ceph orch host` command. You can also have a general view of how the daemons are co-located within a given host, as described in the Red Hat Knowledgebase article https://access.redhat.com/articles/1548993[Red Hat Ceph Storage: Supported configurations].
endif::[]
ifeval::["{build}" != "downstream"]
This ensures that you can visualize the status of the cluster and where daemons are placed by using the `ceph orch host` command, and have a general view of how the daemons are co-located within a given host.
endif::[]

> For this procedure, your environment must meet the following requirements:

Reviewer: ```suggestion
.Prerequisites
```

= Migrating {Ceph} MDS to new nodes within the existing cluster

During the data plane adoption, where the {rhos_prev_long} ({OpenStackShort}) services are redeployed in {rhocp_long}, a {OpenStackPreviousInstaller}-deployed {CephCluster} cluster is migrated in a process called externalizing the {Ceph} cluster.

There are two deployment topologies that currently include an internal {CephCluster} cluster: 

* {OpenStackShort} includes dedicated {CephCluster} nodes to host object storage daemons (OSDs). 
* Hyperconverged Infrastructure (HCI), where Compute nodes double up as {CephCluster} nodes. 

> In either scenario, there are some {Ceph} processes that are deployed on {OpenStackShort} Controller nodes: {Ceph} monitors, Ceph Object Gateway (RGW), Rados Block Device (RBD), Ceph Metadata Server (MDS), Ceph Dashboard, and NFS Ganesha.

Reviewer 2: I think L5 to L12 is more a general introduction. It represents a leftover because initially we wrote these procedures as standalone documents. Now that we connected everything together, I think this introduction can go away entirely.

     ceph osd blocklist rm $item;
done
----
+
[NOTE]
====
When a file system client is unresponsive or misbehaving, the access to the file system might be forcibly terminated. This process is called eviction. Evicting a CephFS client prevents it from communicating further with MDS daemons and OSD daemons.

> Ordinarily, a blocklisted client might not reconnect to the servers; you must unmount and then remount the client. However, in some situations, it might be useful to permit a client that was evicted to attempt to reconnect. Because CephFS uses the RADOS OSD blocklist to control client eviction, you can permit CephFS clients to reconnect by removing them from the blocklist.

Reviewer: ```suggestion
Ordinarily, a blocklisted client cannot reconnect to the servers; you must unmount and then remount the client. However, permitting a client that was evicted to attempt to reconnect can be useful. Because CephFS uses the RADOS OSD blocklist to control client eviction, you can permit CephFS clients to reconnect by removing them from the blocklist.
```

Attempting to remove some redundancies

  label: mds
----

. Use the `ceph orchestrator` to apply the new MDS spec:
+
----
$ sudo cephadm shell -m mds.yaml -- ceph orch apply -i /mnt/mds.yaml
Scheduling new mds deployment …
----
+
This results in an increased number of MDS daemons.

> . Check the new standby daemons that are temporarily added to the CephFS:

Author: Is it correct to just say “CephFS” here?

Reviewer 2: technically "fs" stands for filesystem, but just leaving `CephFS` gives the idea: I think `fs` might result redundant, so I'm ok with your rewriting.

> . To migrate MDS to the right nodes, set the MDS affinity that manages the MDS failover:

Author: What are considered the “right nodes”? Are they the nodes that the user selected to be the target nodes?

Reviewer 2: Yeah, +1 to change to `target nodes` and keep consistency.

ifeval::["{build}" != "downstream"]
[NOTE]
It is possible to elect as "active" a dedicated MDS for a particular file system. To configure this preference, `CephFS` provides a configuration option for MDS called `mds_join_fs`, which enforces this affinity.
When failing over MDS daemons, a cluster’s monitors prefer standby daemons with `mds_join_fs` equal to the file system name with the failed rank. If no standby exists with `mds_join_fs` equal to the file system name, it chooses an unqualified standby as a replacement.
endif::[]
+
----
$ ceph config set mds.mds.cephstorage-0.fqcshx mds_join_fs cephfs
----

. Remove the labels from the Controller nodes and force the MDS failover to the
target node:
+

> ----

Author: Regarding the sentence that says, "The switch happens behind the scenes, and the new active MDS is the one that you set through the `mds_join_fs` command." 

Should this be conditionalized for upstream? There is a note in this procedure that mentions `mds_join_fs` that is upstream only, and without that context, this sentence could be confusing.

Reviewer 2: I think we can remove the conditional and keep that context downstream as well. We need this command in both cases, so it's good to have context.
